include/sisl/sobject/sobject.hpp

```diff
@@ -0,0 +1,116 @@
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <string>
+#include <functional>
+#include <map>
+#include <set>
+#include <memory>
+#include <shared_mutex>
+#include <vector>
+#include <nlohmann/json.hpp>
+
+#include <sisl/logging/logging.h>
+
+namespace sisl {
+
+struct sobject_id {
+    std::string type;
+    std::string name;
+    bool empty() const { return type.empty() && name.empty(); }
+};
+
+typedef struct status_request {
+    nlohmann::json json;
+    bool do_recurse{false};
+    int verbose_level = 0;
+    std::string obj_type;
+    std::string obj_name;
+    std::vector< std::string > obj_path;
+    int batch_size = 10;
+    std::string next_cursor;
+} status_request;
+
+typedef struct status_response {
+    nlohmann::json json;
+} status_response;
+
+using status_callback_type = std::function< status_response(const status_request&) >;
+class sobject;
+using sobject_ptr = std::shared_ptr< sobject >;
+
+[[maybe_unused]] static bool operator<(const sobject_id& id1, const sobject_id& id2) {
+    return id1.type < id2.type || ((id1.type == id2.type) && (id1.name < id2.name));
+}
+
+// To search using only the type as key.
+[[maybe_unused]] static bool operator<(const sobject_id& id, const std::string& key_type) { return id.type < key_type; }
+
+[[maybe_unused]] static bool operator<(const std::string& key_type, const sobject_id& id) { return key_type < id.type; }
+
+[[maybe_unused]] static status_response status_error(std::string error_str) {
+    status_response response;
+    response.json["error"] = error_str;
+    return response;
+}
+
+// Keeps a heirarchy of modules/subsystems which register their callbacks to be
+// whenever a get status is called from the root or directly.
+class sobject {
+public:
+    sobject(const std::string& obj_type, const std::string& obj_name, status_callback_type cb) :
+            m_id{obj_type, obj_name}, m_status_cb(std::move(cb)) {}
+
+    static sobject_ptr create(const std::string& obj_type, const std::string& obj_name, status_callback_type cb) {
+        return std::make_shared< sobject >(obj_type, obj_name, std::move(cb));
+    }
+
+    // Every subsystem add to the json object using update().
+    status_response run_callback(const status_request& request) const;
+    sobject_ptr get_child(const std::string& name);
+    void add_child(const sobject_ptr child);
+
+    sobject_id id() const { return m_id; }
+    std::string name() const { return m_id.name; }
+    std::string type() const { return m_id.type; }
+
+private:
+    sobject_id m_id;
+    std::shared_mutex m_mtx;
+    status_callback_type m_status_cb;
+    // Keep a graph of child nodes. Mapping from name to child status object.
+    std::map< sobject_id, sobject_ptr > m_children;
```

Comment:
what is the reason we don't use std::string, which will be the sobject's name as the key?
From how m_children is being accessed, it looks like using name as key will be much more efficient;

Reply:
Each object is identified with type and name. This is useful in case to get all objects of type 'A' in batch. The map used with key {type, name} with sorted form for easy view.